publish release note for v1.12.0 (#5896)
karmada-io/karmada: docs/CHANGELOG/CHANGELOG-1.12.md

This setup allows secure access through a unified, load-balanced API endpoint. Ultimately, this arrangement enhances resilience against data center outages, complies with disaster recovery requirements, and minimizes the risk of service disruptions.

For a detailed description of this feature, see the [Proposal: Support Custom CA Certificate for Karmada Control Plane](https://github.com/karmada-io/karmada/tree/master/docs/proposals/karmada-operator/custom_ca_cert)

(Feature contributors: @jabellard)

### Karmdactl now supports autocompletion

In the daily operation and maintenance of the Karmada system, many Karmadactl commands are frequently used. 

With this release, Karmadactl now supports autocompletion, which reduces typing significantly and enhances both efficiency and user experience in operating and maintaining Karmadactl. 

Currently, Karmadactl supports command completion for both Bash and Zsh. More information about `karmadactl completion` can be obtained using `karmadactl completion --help`.

(Feature contributors: @zhzhuang-zju)

> 

Reviewer: Add two more new feature describetion:

### Support Roll Back Migration Safely

This release introduced `PreserveResourcesOnDeletion` field to both PropagationPolicy and ClusterPropagationPolicy API, which provides the ability to rollback migration safely. 

The `PreserveResourcesOnDeletion` filed controls whether resources should be preserved on the member clusters when the resource template is deleted in the Karmada control-plane. If it set to true, resources will be preserved on the member clusters.

This feature is particularly useful during workload migration scenarios to ensure that rollback can occur quickly without affecting the workloads running on the member clusters.

For a detailed description of this feature, please refer to [How to Roll Back Migration Operations](https://karmada.io/docs/next/administrator/migration/migrate-in-batch/#how-to-roll-back-migration-operations) and [docs(proposal): Migration Rollback Protection](https://github.com/karmada-io/karmada/pull/5101). 

(Feature contributors: @CharlesQQ @XiShanYongYe-Chang @RainbowMango @a7i @wulemao)

### Stateful Application Failover Supprt

This release introduces a new feature for stateful application failover, it provides a generalized way for users to define application state preservation in the context of cluster-to-cluster failovers.

In the previous releases, Karmada’s scheduling logic runs on the assumption that resources that are scheduled and rescheduled are stateless. In some cases, users may desire to conserve a certain state so that applications can resume from where they left off in the previous cluster.

For CRDs dealing with data-processing (such as Flink or Spark), it can be particularly useful to restart applications from a previous checkpoint. That way applications can seamlessly resume processing data while avoiding double processing.

For a detailed description of this feature, please refer to [[Feature] Stateful Application Failover Support](https://github.com/karmada-io/karmada/issues/5788).

(Feature contributors: @Dyex719 @mszacillo @RainbowMango @XiShanYongYe-Chang)

- API change: The ServiceType of Karmada API server now has been restrict to `ClusterIP`, `NodePort` and `LoadBalancer`. ([#5769](https://github.com/karmada-io/karmada/pull/5581), @RainbowMango)
- Introduced a new condition `CompleteAPIEnablements` to represent api collection status of clusters. ([#5400](https://github.com/karmada-io/karmada/pull/5400), @whitewindmills)
- Introduced `PreserveResourcesOnDeletion` field to both PropagationPolicy and ClusterPropagationPolicy API, which provides the ability to roll back migration safely. ([#5575](https://github.com/karmada-io/karmada/pull/5575), @RainbowMango)
- API Change: Introduced `FieldOverrider` to both OverridePolicy and ClusterOverridePolicy, which provides the ability to override structured data nested in manifest like ConfigMap or Secret. ([#5581](https://github.com/karmada-io/karmada/pull/5581), @RainbowMango)
- Introduced `PurgeMode` to GracefulEvictionTask in ResourceBinding. ([#5816](https://github.com/karmada-io/karmada/pull/5816), @mszacillo)

> - Introduced `StatePreservation` to `PropagationPolicy`, which will be used to preserve status in case of application failover. ([#5885](https://github.com/karmada-io/karmada/pull/5885), @RainbowMango)

Reviewer: > `PropagationPolicy`

Does the CRD resource name in other places need to be unified if quotation marks are not added?

Author: Thanks for pointing this out, I think quotation marks could be added uniformly

### Deprecation
- `ExternalEtcd.CAData`, `ExternalEtcd.CertData` and `ExternalEtcd.KeyData` in `Karmada` API are deprecated and will be removed in a future version. Use SecretRef for providing client connection credentials. ([#5699](https://github.com/karmada-io/karmada/pull/5699), @jabellard)
- The following flags have been deprecated from release `v1.11.0` and now have been removed:
    * `karmada-agent`: ([#5548](https://github.com/karmada-io/karmada/pull/5548), @whitewindmills)
      --bind-address

>       --secure-port

Reviewer: This effect does not seem to be expected:

![image](https://github.com/user-attachments/assets/be01a1fb-df42-4728-ac54-1594dd5240c2)

### Features & Enhancements
- `karmada-controller-manager`: introduces the `agentcsrapproving` controller to provide the capability for the agent's CSR to be automatically approved. ([#5825](https://github.com/karmada-io/karmada/pull/5825), @zhzhuang-zju)
- `karmada-controller-manager`: update taint-manager to config eviction task with purgeMode. ([#5879](https://github.com/karmada-io/karmada/pull/5879), @XiShanYongYe-Chang)
- `karmada-controller-manager`: Build eviction task for application failover when using purgeMode Immediately. ([#5881](https://github.com/karmada-io/karmada/pull/5881), @mszacillo)
- `karmada-controller-manager`: build PreservedLabelState when triggering evition in RB/CRB application controller. ([#5887](https://github.com/karmada-io/karmada/pull/5887), @XiShanYongYe-Chang)
- `karmada-controller-manager`: keep preserveResourcesOnDeletion of the dependent resource consistent with that of the primary resource. ([#5717](https://github.com/karmada-io/karmada/pull/5717), @XiShanYongYe-Chang)

> - `karmada-controller-manager`: set conflictResolution for dependent resources. ([#4418](https://github.com/karmada-io/karmada/pull/4418), @@chaunceyjiang)

Reviewer: @@chaunceyjiang -> @chaunceyjiang

For a detailed description of this feature, please refer to [How to Roll Back Migration Operations](https://karmada.io/docs/next/administrator/migration/migrate-in-batch/#how-to-roll-back-migration-operations) and [docs(proposal): Migration Rollback Protection](https://github.com/karmada-io/karmada/pull/5101).

(Feature contributors: @CharlesQQ @XiShanYongYe-Chang @RainbowMango @a7i @chaosi-zju @wulemao)

### Stateful Application Failover Support

This release introduces a new feature for stateful application failover, it provides a generalized way for users to define application state preservation in the context of cluster-to-cluster failovers.

In the previous releases, Karmada’s scheduling logic runs on the assumption that resources that are scheduled and rescheduled are stateless. In some cases, users may desire to conserve a certain state so that applications can resume from where they left off in the previous cluster.

For CRDs dealing with data-processing (such as Flink or Spark), it can be particularly useful to restart applications from a previous checkpoint. That way applications can seamlessly resume processing data while avoiding double processing.

> For a detailed description of this feature, please refer to [[Feature] Stateful Application Failover Support](https://github.com/karmada-io/karmada/issues/5788).

Reviewer 2: The referencing issue does not provide much useful context for this feature.
@XiShanYongYe-Chang Are you writing the docs of this feature?

Reviewer: Writing in progress

Reviewer: It's ready now: https://github.com/karmada-io/website/pull/749

Reviewer: https://karmada.io/docs/next/userguide/failover/application-failover/#stateful-application-failover-support

(Feature contributors: @Dyex719 @mszacillo @RainbowMango @XiShanYongYe-Chang)

### Karmada Operator Enhancement: Multi-Data Center Redundancy and Disaster Recovery(DR)

This release, karmada-operator introduces support for a Custom CA Certificate for Karmada Instances and the ability to utilize external etcd. 
These enhancements allow Karmada-operator to deploy a highly available managed Karmada control plane across multiple management clusters that can span various data centers, thus fulfilling disaster recovery requirements.

By implementing this architecture and configuring the managed control plane instances to use the same CA certificate and underlying etcd instance, you can create a stretched instance that operates across multiple management clusters. 

> This setup allows secure access through a unified, load-balanced API endpoint. Ultimately, this arrangement enhances resilience against data center outages, complies with disaster recovery requirements, and minimizes the risk of service disruptions.

Reviewer 2: We enhanced karmada-operator a lot in this release, except this one also includes:
- #5242 
- #5257 
- #5767

- Introduced `SecretRef` to `Karmada` API as part of the configuration for connecting to an external etcd cluster can be used to reference a secret that contains credentials for connecting to an external etcd cluster. ([#5699](https://github.com/karmada-io/karmada/pull/5699), @jabellard)
- Introduced `extraVolumes` and `extraVolumemounts` to the `Karmada` API to optionally specify extra volumes and volume mounts for the Karmada API server component. ([#5509](https://github.com/karmada-io/karmada/pull/5509), @jabellard)
- Introduced `ApiServerService` field to `Karmada` API as part of the Karmada instance status can be used to reference the API Server service for that instance. This is useful for scenarios where higher level operators need to discover the API Server service of a Karmada instance  for tasks like setting up ingress traffic. ([#5775](https://github.com/karmada-io/karmada/pull/5775), @jabellard)
- Introduced `CustomCertificate.ApiServerCACert` field to `Karmada` API as part of the `Karmada` spec to specify the reference to a secret that contains a custom CA certificate for the Karmada API Server. ([#5842](https://github.com/karmada-io/karmada/pull/5842), @jabellard)

> - API change: The `ServiceType` of Karmada API server now has been restrict to `ClusterIP`, `NodePort` and `LoadBalancer`. ([#5769](https://github.com/karmada-io/karmada/pull/5581), @RainbowMango)

Reviewer 2: Incorrect link.

```suggestion
- The ServiceType of the Karmada API server in `Karmada` API now has been restricted to `ClusterIP`, `NodePort`, and `LoadBalancer`. ([#5769](https://github.com/karmada-io/karmada/pull/5769), @RainbowMango)
```

> - Introduced a new condition `CompleteAPIEnablements` to represent api collection status of clusters. ([#5400](https://github.com/karmada-io/karmada/pull/5400), @whitewindmills)

Reviewer 2: ```suggestion
- Introduced a new condition `CompleteAPIEnablements` to `Cluster` API to represent the API collection status. ([#5400](https://github.com/karmada-io/karmada/pull/5400), @whitewindmills)
```
I also updated the release notes on that PR.

- Introduced `PreserveResourcesOnDeletion` field to both `PropagationPolicy` and `ClusterPropagationPolicy` API, which provides the ability to roll back migration safely. ([#5575](https://github.com/karmada-io/karmada/pull/5575), @RainbowMango)
- API Change: Introduced `FieldOverrider` to both `OverridePolicy` and `ClusterOverridePolicy`, which provides the ability to override structured data nested in manifest like ConfigMap or Secret. ([#5581](https://github.com/karmada-io/karmada/pull/5581), @RainbowMango)

> - Introduced `PurgeMode` to `GracefulEvictionTask` in ResourceBinding. ([#5816](https://github.com/karmada-io/karmada/pull/5816), @mszacillo)

Reviewer 2: ```suggestion
- Introduced `PurgeMode` to `GracefulEvictionTask` in `ResourceBinding` and `ClusterResourceBinding` API. ([#5816](https://github.com/karmada-io/karmada/pull/5816), @mszacillo)
```

- Introduced `StatePreservation` to `PropagationPolicy`, which will be used to preserve status in case of application failover. ([#5885](https://github.com/karmada-io/karmada/pull/5885), @RainbowMango)

### Deprecation
- `ExternalEtcd.CAData`, `ExternalEtcd.CertData` and `ExternalEtcd.KeyData` in `Karmada` API are deprecated and will be removed in a future version. Use SecretRef for providing client connection credentials. ([#5699](https://github.com/karmada-io/karmada/pull/5699), @jabellard)

> - The following flags have been deprecated from release `v1.11.0` and now have been removed:

Reviewer 2: ```suggestion
- The following flags have been deprecated in release `v1.11.0` and now have been removed:
```

(Feature contributors: @CharlesQQ @XiShanYongYe-Chang @RainbowMango @a7i @chaosi-zju @wulemao)

### Stateful Application Failover Support

This release introduces a new feature for stateful application failover, it provides a generalized way for users to define application state preservation in the context of cluster-to-cluster failovers.

In the previous releases, Karmada’s scheduling logic runs on the assumption that resources that are scheduled and rescheduled are stateless. In some cases, users may desire to conserve a certain state so that applications can resume from where they left off in the previous cluster.

For CRDs dealing with data-processing (such as Flink or Spark), it can be particularly useful to restart applications from a previous checkpoint. That way applications can seamlessly resume processing data while avoiding double processing.

For a detailed description of this feature, please refer to [[Feature] Stateful Application Failover Support](https://github.com/karmada-io/karmada/issues/5788).

> (Feature contributors: @Dyex719 @mszacillo @Rai

Reviewer: I'm missing the right parenthesis.

- `karmadactl`: Implementing autocompletion for karmadactl to save a lot of typing. ([#5533](https://github.com/karmada-io/karmada/pull/5533), @zhzhuang-zju)
- `karmadactl`: Added shorthand letter `s` to 'operation-scope' flags across commands. ([#5483](https://github.com/karmada-io/karmada/pull/5483), @ahorine)
- `karmadactl`: `karmadactl init` support multiple label selection ability with flag `EtcdNodeSelectorLabels`. ([#5321](https://github.com/karmada-io/karmada/pull/5321), @tiansuo114)
- `karmadactl`: `karmadactl init` supports deployment through configuration files. ([#5357](https://github.com/karmada-io/karmada/pull/5357), @tiansuo114)
- `karmadactl`: new command `karmadactl unregister` supports unregister a pull mode cluster. ([#5626](https://github.com/karmada-io/karmada/pull/5626), @wulemao)
- `karmadactl`: set `PreserveResourcesOnDeletion` by default in auto-created propagation policy during promotion process. ([#5601](https://github.com/karmada-io/karmada/pull/5601), #wulemao)
- `karmadactl`: The `--force` option of `unjoin` command now try to clean up resources propagated in member clusters. ([#4451](https://github.com/karmada-io/karmada/pull/4451), @zhzhuang-zju)

> - `karmadactl`: command `register` minimizes the RBAC permissions for the registering cluster to access Karmada control plane. ([#5793](https://github.com/karmada-io/karmada/pull/5793), @zhzhuang-zju)

Author: It may be inaccurate because it is the cluster, not the command, that has restricted access permission.

> - `karmadactl`: command `register` minimizes the RBAC permissions for the registering cluster to access Karmada control plane. ([#5793](https://github.com/karmada-io/karmada/pull/5793), @zhzhuang-zju)

Author: How about:
```suggestion
- `karmadactl`: RBAC permissions for pull mode clusters registered with the `register` command are minimized when accessing the Karmada control plane. ([#5793](https://github.com/karmada-io/karmada/pull/5793), @zhzhuang-zju)
```

> - `karmadactl`: command `register` minimizes the RBAC permissions for the registering cluster to access Karmada control plane. ([#5793](https://github.com/karmada-io/karmada/pull/5793), @zhzhuang-zju)

Author: > better than previous

Yeah, agree~
